rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/operation/RecurseDirectoryAndAddFiles.java

```diff
@@ -56,8 +63,13 @@ private void recurseAndAddFiles(FileService fileService, FileModel file)
             {
                 for (File reference : subFiles)
                 {
+                    if (FileDiscoveredListenerUtil.shouldSkip(listeners, new DefaultFileDiscoveredEvent(reference.getAbsolutePath())))
+                        continue;
+
                     FileModel subFile = fileService.createByFilePath(file, reference.getAbsolutePath());
-                    recurseAndAddFiles(fileService, subFile);
+                    for (FileDiscoveredListener listener : listeners)
+                        listener.fileModelCreated(event, context, subFile);
```

Comment:
Q: Why do we use listeners in case the rule may see all the created files just by querying FileModels in the given phase?